src/providers/info.rs

```diff
@@ -139,6 +146,87 @@ impl InfoProvider for DefaultInfoProvider<'_> {
 }
 
 impl DefaultInfoProvider<'_> {
+    pub fn new<'p>(context: &'p Context) -> DefaultInfoProvider<'p> {
+        DefaultInfoProvider {
+            client: context.client(),
+            cache: context.cache(),
+            safe_cache: HashMap::new(),
+            token_cache: HashMap::new(),
+        }
+    }
+
+    fn cached<T>(
+        &mut self,
+        local_cache: impl Fn(&mut Self) -> &mut HashMap<String, Option<T>>,
+        generator: impl Fn(&mut Self, &String) -> Result<Option<T>>,
+        key: impl Into<String>,
+    ) -> Result<T>
+    where
+        T: Clone + DeserializeOwned,
+    {
+        let key = key.into();
+        match local_cache(self).get(&key) {
+            Some(value) => value
+                .clone()
+                .ok_or(anyhow::anyhow!("Cached value not available")),
+            None => {
+                let value: Option<T> = generator(self, &key)?;
+                local_cache(self).insert(key, value.clone());
+                value.ok_or(anyhow::anyhow!("Could not generate value"))
+            }
+        }
+    }
+
+    fn load_safe_info(&mut self, safe: &String) -> Result<Option<SafeInfo>> {
+        let url = format!("{}/v1/safes/{}/", base_transaction_service_url(), safe);
+        let data: String = self
+            .cache
+            .request_cached(self.client, &url, info_cache_duration())?;
+        Ok(serde_json::from_str(&data).unwrap_or(None))
+    }
+
+    fn populate_token_cache(&mut self) -> Result<()> {
+        let url = format!("{}/v1/tokens/?limit=10000", base_transaction_service_url());
+        let response = self.client.get(&url).send()?;
+        let data: Page<TokenInfo> = response.json()?;
+        for token in data.results.iter() {
+            self.cache.create(
+                &format!("dip_ti_{}", token.address),
+                &serde_json::to_string(&token)?,
+                info_cache_duration(),
+            )
+        }
+        Ok(())
+    }
+
+    fn check_token_cache(&mut self) -> Result<()> {
+        if self.cache.fetch("dip_tcl").is_some() {
+            // Cache is still up to data
+            return Ok(());
+        }
+        let result = self.populate_token_cache();
+        // If error we use a shorter cache timeout (we do not want to DoS our service in case of an error)
+        self.cache.create(
+            "dip_tcl",
+            "",
+            if result.is_ok() {
+                info_cache_duration()
+            } else {
+                10000
+            },
+        );
+        result
```

Comment:
@rmeissner , I am not sure what's the purpose of `dip_tcl`, why not cache the result of `populate_token_Cache` request with the usual timeout, like we do? Is this so that we check regularly for the full list, but the individual items are checked rarely? 

I was thinking we could add a new payload to trigger this from the `/v1/hook/update/<token>` endpoint. Then we could have a crontab somewhere issuing the request.

Reply:
because this would mean we cache the whole data twice and as it is quite some data I would avoid this.